### ISSUE TITLE
BleachBit creates a bunch of random files at the end of wiping a large drive on Windows.

### ISSUE THREAD

**Issue by StaticallyTypedRice** [bleachbit/bleachbit#502]
# Issue:

After wiping a large drive on Windows, BleachBit tends to create (sometimes hundreds) of files with random file names in the root folder of the drive with no content, then returns an error that one of the files is "being used by another process". When BleachBit is closed after this, all but one of these files are automatically deleted.

This problem occurs on Bleachbit version 2.0, though I have not tested newer versions.

# Steps to Reporduce

Install BleachBit on Windows and wipe a large drive. I've seen this happen every time on drives of several hundred gigabytes, but it may also happen on smaller drives.

Wait until the end of the wiping process, and look at the root folder (or wherever the wipe file is) before closing BleachBit.

**Comment by az0** [comment]
Always one file is left? How big is the file? Have you tried this when running `bleachbit_console.exe`? This program sometimes shows more information.

BleachBit version 2.2 is out now, but I expect it may be the same in this regard.

**Comment by StaticallyTypedRice** [comment]
> Always one file is left? How big is the file? Have you tried this when running `bleachbit_console.exe`? This program sometimes shows more information.
> 
> BleachBit version 2.2 is out now, but I expect it may be the same in this regard.

Yes, always. I assume it's the file that BleachBit was complaining about. It says it can't close it. It's usually empty (as are all but the main file BleachBit uses to fill the file system with zeros), but all the random files have the same naming structure as the main wipe file, starting with "tmp" followed by a bunch of random characters.

It's been a while since I've wipe any hard drives, but I'll try doing some wipes later so I can more carefully observe the behaviour.

**Comment by az0** [comment]
Those files sound like the kind that BleachBit makes while it's working.

You mentioned "it says it can't close." Does this mean you terminated the BleachBit process?

**Comment by StaticallyTypedRice** [comment]
I remember seeing an error along the lines of "bleachbit can't close this file because a process is using it" or maybe it was "bleachbit can't close this file because a process is using it" (now that I think about it, the latter makes more sense. I wasn't sure, so I edited that part out.

**Comment by StaticallyTypedRice** [comment]
Update: So I did some tests with bleachbit 2.2, and here are the results.

When I tested with a removable disk, Bleachbit successfully wiped it without issue. Though I think It was still creating those files at the end of the wiping process, but was just able to delete them all. I think this because I once saw it creating those files when I opened the drive I was wiping in WIndows explorer toward the end of the wipe.

When I tested with the boot disk of my windows computer though, it showed this error:

```
[Error 32] The process cannot access the file because it is being used by another process: 'C:\\tmp60_xzslT17rJp7Wq4xk4i2gtAN9KtlbK3iPV36CAEXEogJHb30gXLc65KHI4rp2pUiObi-1UrhCHlCWNWGy..xOsT._CgvTwKU8lYc_sMZldH7sfAummizkVsCrh6ej9PxsKIC7Tq_ZLbEMkH-p9Kfg-R9BNSYC1x4-gdOR7oQCPHiSBp89Tv42JcSGnpm3uNYILEYEZg1EL4CB8ljh7ETc1eO3UE23o2ipE7NRH8oYaFCo8u2oz.JmKeif': Function: Overwrite free disk space C:\

Disk space recovered: 0
Files deleted: 0
Errors: 1
```

Listing the directory of the C drive before closing the program, we get this:

```PS C:\> dir


    Directory: C:\


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       2019-03-21   3:48 PM                Microsoft
d-----       2018-09-15  12:33 AM                PerfLogs
d-r---       2019-03-23   5:10 PM                Program Files
d-r---       2019-03-21   3:39 PM                Program Files (x86)
d-r---       2018-12-09   2:00 AM                Users
d-----       2019-03-25   6:00 PM                Windows
d-----       2018-11-21   5:09 PM                Windows10Upgrade
-a----       2019-03-25   5:55 PM              1 tmp0bo1fweKzbEYUVbZpNuPNEidx5kU2EahrVGts3YwUABwvTmWL97mlvSPEtvhcgpLsrn
                                                 BzLhJsuNi6fjlh2okHB4pd57SNWLURFO20RG1.LGUr0WXn5eP7fD00ctdIFWL1R4FM4ak7
                                                 vDdd2uDpjZHzPCBfHy1ybU0md.07Ly8iTtNtBg07_oKameFbSdsvutFcEvBjfQv6UcbKmm
                                                 TJ_qwrhnE5rMJB.rZgEuyPUkXnGiLQ6kTclMe8h-bEnd
-a----       2019-03-25   5:55 PM              1 tmp0ig_1buzykFsu49fRB8YLyAf.4Ue28uusRQI8wdqnMBDRKUDElzXFvk6KgB8GBQuyu3
                                                 MZ37IOOmLFw31wgE7jkBx_2a6uAuo2B_v_o0FWWb-FgqXR_L_1b..r4nkK5HQDcqcKphek
                                                 OVXu9w_J9mY5qcIywh7d0NqvQx4.Dc1CHt4ixOoPz27FadZ4wXKymSWCuCq4wTZv-zu_I8
                                                 H7BlpKcpZ6T-EsGa14GlkRj6sA0OrA1NnUg8s5nwxL__
-a----       2019-03-25   5:55 PM              1 tmp0mbjakLJjF3XznWwyE60LmF.GxyJr56nM5nY4TA-wvj2XRjt5pzmDNwaNTmFEc9Iapn
                                                 Xyxi3ompEfuCM_2eTGJIU1w7XJURGLLQIgJIIxkwmF3.19_PQTYbFoIOQZ3-57oCTJnYB1
                                                 j9rlEjdnOLdFCrP-1g6dCZOJwrG-NBmfIOqxKYlkG3TZTSnpU.obgx7.0E7eyzR2ai9Zjs
                                                 -xpFI0l8GB0oydh__8zMKMVkmpGzyD9qbKGSHbJsJ_Y5
<There are hundreds more files...>

-a----       2019-03-25   5:55 PM         253952 tmpzt65xkY5MWtFn2ol5c3rQ45JYz5f0K1STey5Lxq3wkxL.vBSTB5PCPfIB8yeUNLMUkH
                                                 I2MEyGVTp_GhFZjFLf3xy8Wmjj8Tt_XhLfQJnBzJ17KAWWawIQ8VRNndTq8QqRelnwRDCp
                                                 rsc5CMGu35yZnyZUlJeqBTr5PRc0JqyVt2Bbl6LmLoxP38pUU7ifshxGcl333ogx25bu09
                                                 NTVv7FUtfDFDkG3B7dj4oTz5w-CgGNLnmcoaAWgps9PA
						
						<I think this is the main wipe file. It is the only one with data in it.>

<Even more files after this...>
```

After closing the program, only one file remains, like I previously said. Only difference is that the file that remains is empty while these files have a length of one. I regrettably didn't open the files in a hex editor to see what they contained, but my guess they might just contain a single null character.

There were no entries about this error in any BleachBit log files that I could find.

**Comment by az0** [comment]
I put in a change that might help. Would you please test BleachBit version 2.3.1114 or later from the [CI site](https://ci.bleachbit.org/)?

You could test it normally like you did, or for extra information try the command below from a console with administrative privileges. (If you want more instructions on how to run this command, let me know.)

`c:\Program Files (x86)\BleachBit\bleachbit_console.exe --debug-log=c:\ --wipe-free-space c:\`

**Comment by mkhon** [comment]
On Windows 10 I see that after ENOSPC f.tell() returns 8192 so the following code does not break the loop:

```
        # If no bytes were written, then quit.
        # See https://github.com/bleachbit/bleachbit/issues/502
        if f.tell() < 2:
            break
```

**Comment by mkhon** [comment]
Fixed in #833

**Comment by az0** [comment]
@StaticallyTypedRice 

Would you please test Max's fix? In a few minutes, there will be a 3.9.1.x build available on the [CI site](https://ci.bleachbit.org/). Please test and let us know whether it helped.

**Comment by az0** [comment]
@StaticallyTypedRice 

The latest package is available here for testing:[ BleachBit 3.9.2 beta](https://www.bleachbit.org/news/bleachbit-392-beta). After testing, please add a comment here.

**Comment by az0** [comment]
I am closing because we made a code fix. 

@StaticallyTypedRice if the exact same issue remains, please reopen this issue. If you have a related problem or different problem, open a new issue.
